examples/bcc/execsnoop/execsnoop.go

```diff
@@ -0,0 +1,308 @@
+// from the current task, so we need to use this fallback to have
+// the parent PID in any kernel.
+func getPpid(pid uint64) uint64 {
+	f, err := os.OpenFile(fmt.Sprintf("/proc/%d/status", pid), os.O_RDONLY, os.ModePerm)
+	if err != nil {
+		return 0
+	}
+	defer f.Close()
+
+	sc := bufio.NewScanner(f)
+	for sc.Scan() {
+		text := sc.Text()
+		if strings.Contains(text, "PPid:") {
+			f := strings.Fields(text)
+			i, _ := strconv.ParseUint(f[len(f)-1], 10, 64)
+			return i
+		}
+	}
+	return 0
+}
+
+func main() {
+	traceFailed := flag.Bool("x", false, "trace failed exec()s")
+	timestamps := flag.Bool("t", false, "include timestamps")
+	quotemarks := flag.Bool("q", false, `add "quotemarks" around arguments`)
+	filterComm := flag.String("n", "", `only print command lines containing a name, for example "main"`)
+	filterArg := flag.String("l", "", `only print command where arguments contain an argument, for example "tpkg"`)
+	format := flag.String("o", "table", "output format, either table or json")
+	pretty := flag.Bool("p", false, "pretty print json output")
+	maxArgs := flag.Uint64("m", 20, "maximum number of arguments parsed and displayed, defaults to 20")
+
+	flag.Parse()
+
+	m := bpf.NewModule(strings.Replace(source, "MAX_ARGS", strconv.FormatUint(*maxArgs, 10), -1), []string{})
+	defer m.Close()
+
+	kprobe, err := m.LoadKprobe("syscall__execve")
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to load syscall__execve: %s\n", err)
+		os.Exit(1)
+	}
+	if err := m.AttachKprobe("sys_execve", kprobe); err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to attach syscall__execve: %s\n", err)
+		os.Exit(1)
+	}
+
+	kretprobe, err := m.LoadKprobe("do_ret_sys_execve")
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to load do_ret_sys_execve: %s\n", err)
+		os.Exit(1)
+	}
+	if err := m.AttachKretprobe("sys_execve", kretprobe); err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to attach do_ret_sys_execve: %s\n", err)
+		os.Exit(1)
+	}
+
+	table := bpf.NewTable(m.TableId("events"), m)
+
+	channel := make(chan []byte)
+
+	perfMap, err := bpf.InitPerfMap(table, channel)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to init perf map: %s\n", err)
+		os.Exit(1)
+	}
+
+	sig := make(chan os.Signal, 1)
+	signal.Notify(sig, os.Interrupt, os.Kill)
+
+	args := make(map[uint64][]string)
+
+	out := newOutput(*format, *pretty, *timestamps)
+	out.PrintHeader()
+
+	go func() {
+		var event execveEvent
+		for {
+			data := <-channel
+			err := binary.Read(bytes.NewBuffer(data), binary.LittleEndian, &event)
+			if err != nil {
+				fmt.Printf("failed to decode received data: %s\n", err)
+				continue
+			}
+			if eventArg == EventType(event.Type) {
+				e, ok := args[event.Pid]
+				if !ok {
+					e = make([]string, 0)
+				}
+				argv := (*C.char)(unsafe.Pointer(&event.Argv))
+
+				e = append(e, C.GoString(argv))
+				args[event.Pid] = e
+			} else {
+				if event.RetVal != 0 && !*traceFailed {
+					continue
```

Comment:
Following the implementation in bcc, you also have to clear `args` in the continue cases I think (this line and three below).

Otherwise arguments get mixed up and repeated. On my system I see for example:

```
ifconfig         32250  755      0 br-ad8b079223dd /usr/local/sbin/ifconfig /usr/local/bin/ifconfig br-ad8b079223dd /usr/sbin/ifconfig br-ad8b079223dd /usr/bin/ifconfig br-ad 8b079223dd /sbin/ifconfig br-ad8b079223dd
```